types: match MySQL behavior with datetime delimiters

smola opened the pull request:
### What problem does this PR solve?

Issue Number: close #17277

Problem Summary: `select date '2020-02--1'` fails 

### What is changed and how it works?

* `types.ParseDateFormat` changed datetime splitting to match MySQL 5.7.
* Any ASCII punctuation character (i.e. printable non-alphanumeric) is a valid delimiter.
* Additionally, ` ` (space) and `T` are valid delimiters if they appear between time and date.
* Multiple consecutive delimiters are accepted, and handled as a single delimiter.
* Consecutive delimiters can be mixed (e.g. `-./` or ` T/TT `).
* This PR does _not_ apply to `types.ParseDuration`, since it has additional validation.

### Check List

- Unit tests

#### Backwards compatibility

* In previous TiDB versions, `T` and ` ` (space) were valid delimiters anywhere in a datetime string. After this change, using `T` or ` ` (space) as a delimiter is an error except between date and time. 

### Release note

* Datetime parsing now matches MySQL 5.7 behavior for delimiters.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/pingcap/tidb?pullRequest=17376) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/pingcap/tidb?pullRequest=17376) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/pingcap/tidb?pullRequest=17376) it.</sub>

Reminiscent commented:
The unit-test is failed. Please fix them. Thanks!
```
[2020-05-25T07:44:53.516Z] FAIL: integration_test.go:1174: testIntegrationSuite2.TestTimeBuiltin

[2020-05-25T07:44:53.516Z] 

[2020-05-25T07:44:53.516Z] integration_test.go:1446:

[2020-05-25T07:44:53.516Z]     result = tk.MustQuery("select addtime(a, b), addtime(cast(a as date), b), addtime(b,a), addtime(a,c), addtime(b," +

[2020-05-25T07:44:53.516Z]         "c), addtime(c,a), addtime(c,b)" +

[2020-05-25T07:44:53.516Z]         " from t;")

[2020-05-25T07:44:53.516Z] /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:288:

[2020-05-25T07:44:53.516Z]     tk.c.Check(err, check.IsNil, comment)

[2020-05-25T07:44:53.516Z] ... value *errors.withStack = [types:1292]Incorrect time value: '{0 0 0 1 1 1 0}' ("[types:1292]Incorrect time value: '{0 0 0 1 1 1 0}'")

[2020-05-25T07:44:53.516Z] ... sql:select addtime(a, b), addtime(cast(a as date), b), addtime(b,a), addtime(a,c), addtime(b,c), addtime(c,a), addtime(c,b) from t;, args:[]

[2020-05-25T07:44:53.516Z] 

[2020-05-25T07:44:53.516Z] integration_test.go:1449:

[2020-05-25T07:44:53.516Z]     result.Check(testkit.Rows("<nil> <nil> <nil> 2017-01-01 13:31:32 2017-01-01 13:31:32 <nil> <nil>"))

[2020-05-25T07:44:53.516Z] /home/jenkins/agent/workspace/tidb_ghpr_check_2/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:62:

[2020-05-25T07:44:53.516Z]     res.c.Assert(resBuff.String(), check.Equals, needBuff.String(), res.comment)

[2020-05-25T07:44:53.516Z] ... obtained string = ""

[2020-05-25T07:44:53.516Z] ... expected string = "[<nil> <nil> <nil> 2017-01-01 13:31:32 2017-01-01 13:31:32 <nil> <nil>]\n"

[2020-05-25T07:44:53.516Z] ... sql:select addtime(a, b), addtime(cast(a as date), b), addtime(b,a), addtime(a,c), addtime(b,c), addtime(c,a), addtime(c,b) from t;, args:[]
```

smola commented:
@Reminiscent Oops. I missed that one. I have fixed the integration test, it used `2017 01-01 12:30:31` as a date, which produces an error in MySQL (see the initial notes about backwards compatibility). So I changed the test to `2017-01-01 12:30:31`.

Just for the record, I tested on MySQL 8 with the following:
```sql
create database if not exists test;
use test;
drop table if exists t;
create table t(a datetime, b timestamp, c time);
insert into t values("2017 01-01 12:30:31", "2017 01-01 12:30:31", "01:01:01");
```

This produces:
```
Error Code: 1292. Incorrect datetime value: '2017 01-01 12:30:31' for column 'a' at row 1
```

I still have 4 unit tests failing locally, but they fail on master too. They are probably related to https://github.com/pingcap/tidb/issues/17013

Reminiscent commented:
/merge
